Graph.java

```diff
@@ -0,0 +1,166 @@
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+
+/**
+ * Created by basescu on 06.06.17.
+ */
+
+class Node {
+    public int key;     // letter index in initial matrix: for M[i][j], we have pos = i * cols + j
```

Comment:
final! Especially if field is publicly accessible and thus could be modified by someone outside your control (imagine that this code is part of a bigger project, where more people can use it).
Same for value below.